fendermint/vm/snapshot/src/manager.rs

```diff
@@ -0,0 +1,197 @@
+    ) -> (Self, SnapshotClient) {
+        let manager = Self {
+            client,
+            store,
+            snapshot_dir,
+            snapshot_state: SnapshotState {
+                // Start with nothing to snapshot until we are notified about a new height.
+                // We could also look back to find the latest height we should have snapshotted.
+                latest_params: TVar::new(None),
+            },
+            sync_poll_interval,
+            // Assume we are syncing until we can determine otherwise.
+            is_syncing: TVar::new(true),
+        };
+        let client = SnapshotClient {
+            snapshot_interval,
+            snapshot_state: manager.snapshot_state.clone(),
+        };
+        (manager, client)
+    }
+
+    /// Produce snapshots.
+    pub async fn run(self) {
+        // Start a background poll to CometBFT.
+        // We could just do this once and await here, but this way ostensibly CometBFT could be
+        // restarted without Fendermint and go through another catch up.
+        {
+            let client = self.client.clone();
+            let is_syncing = self.is_syncing.clone();
+            let poll_interval = self.sync_poll_interval;
+            tokio::spawn(async move {
+                poll_sync_status(client, is_syncing, poll_interval).await;
+            });
+        }
+
+        let mut last_params = None;
+        loop {
+            let (params, height) = atomically(|| {
+                // Check the current sync status. We could just query the API, but then we wouldn't
+                // be notified when we finally reach the end, and we'd only snapshot the next height,
+                // not the last one as soon as the chain is caught up.
+                if *self.is_syncing.read()? {
+                    retry()?;
+                }
+
+                match self.snapshot_state.latest_params.read()?.as_ref() {
+                    None => retry()?,
+                    unchanged if *unchanged == last_params => retry()?,
+                    Some(new_params) => Ok(new_params.clone()),
+                }
+            })
+            .await;
+
+            if let Err(e) = self.create_snapshot(height, params.clone()).await {
+                tracing::warn!(error =? e, height, "failed to create snapshot");
+            }
+
+            last_params = Some((params, height));
+        }
+    }
+
+    /// Export a snapshot to a temporary file, then copy it to the snapshot directory.
+    async fn create_snapshot(
+        &self,
+        height: BlockHeight,
+        params: FvmStateParams,
+    ) -> anyhow::Result<()> {
+        let snapshot = Snapshot::new(self.store.clone(), params, height)
+            .context("failed to create snapshot")?;
+
+        let file_name = format!("snapshot-{height}.car");
+        let file_path = self.snapshot_dir.join(file_name.clone());
+        let temp_dir = tempfile::tempdir().context("failed to create temp dir for snapshot")?;
```

Comment:
Why do you need to have an intermediate temp file, in case it fails generating the car?

Reply:
Yes, I thought if the application crashes in the middle of exporting a snapshot, this should make sure the half-finished dump will be cleaned up.